[BasicAA] Scalable offset with scalable typesize.

davemgreen opened the pull request:
This patch adds a simple alias analysis check for accesses that are scalable
with a offset between them that is also trivially scalable (there are no other
constant/variable offsets). We essentially divide each side by vscale and are
left needing to check that the offset >= typesize.

nikic reviewed (commented):
> We essentially divide each side by vscale and are left needing to check that the offset >= typesize.

Is that correct without the IsNSW flag?

davemgreen commented:
> > We essentially divide each side by vscale and are left needing to check that the offset >= typesize.
> 
> Is that correct without the IsNSW flag?

Hmm. Yeah maybe. I was thinking of offsets similar to typesizes, and it would be a strange access than might wrap around depending on the vscale. I'll add a IsNSW check if you think that's better - it looks like a few cases in the tests get a little worse, but we might be able to improve it in the future with vscale_range to show the extreme would not overflow.

davemgreen commented:
Added a check for `DecompGEP1.VarIndices[0].IsNSW`

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1170,6 +1170,26 @@ AliasResult BasicAAResult::aliasGEP(
+  if (DecompGEP1.VarIndices.size() == 1 && DecompGEP1.VarIndices[0].IsNSW &&
+      DecompGEP1.Offset.isZero() &&
+      PatternMatch::match(DecompGEP1.VarIndices[0].Val.V,
+                          PatternMatch::m_VScale())) {
```
I think we also need `DecompGEP1.VarIndices[0].Val.TruncBits == 0`.

nikic reviewed (approved):
LGTM

davemgreen committed:
[BasicAA] Scalable offset with scalable typesize.

This patch adds a simple alias analysis check for accesses that are scalable
with a offset between them that is also trivially scalable (there are no other
constant/variable offsets). We essentially divide each side by vscale and are
left needing to check that the offset >= typesize.
